Xamarin.PropertyEditing.Windows/TreeViewItemEx.cs

```diff
@@ -0,0 +1,264 @@
+
+			IndentLevel = level;
+		}
+
+		protected override void OnMouseDown (MouseButtonEventArgs e)
+		{
+			if (IsSelectable) {
+				if (Keyboard.IsKeyDown (Key.LeftCtrl) || Keyboard.IsKeyDown (Key.RightCtrl)) {
+					IsSelected = !IsSelected;
+					e.Handled = true;
+				}
+			}
+
+			base.OnMouseDown (e);
+		}
+
+		protected override void OnKeyDown (KeyEventArgs e)
+		{
+			if (e.Key == Key.Up || e.Key == Key.Down) {
+				var next = GetNextElement (e.Key == Key.Up) as ItemsControl;
+				if (!(next is TreeView)) {
+					if (next.HasItems) {
+						var tree = GetParentTree();
+						tree.SelectedTreeItem = null;
+					}
+
+					next.Focus();
+					e.Handled = true;
+				}
+			}
+
+			base.OnKeyDown (e);
+		}
+
+		protected override void OnSelected (RoutedEventArgs e)
+		{
+			TreeViewEx parent = GetParentTree();
+			if (!IsSelectable) {
+				IsSelected = false;
+				e.Handled = true;
+
+				parent.SelectedTreeItem = parent.SelectedTreeItem;
+			} else if (parent.SelectedTreeItem != this) {
+				parent.SelectedTreeItem = this;
+			}
+		}
+
+		protected override DependencyObject GetContainerForItemOverride ()
+		{
+			return new TreeViewItemEx();
+		}
+
+		protected override bool IsItemItsOwnContainerOverride (object item)
+		{
+			return item is TreeViewItemEx;
+		}
+
+		protected override void OnMouseEnter (MouseEventArgs e)
+		{
+			base.OnMouseEnter (e);
+
+			CheckPosition (e);
+		}
+
+		protected override void OnMouseMove (MouseEventArgs e)
+		{
+			base.OnMouseMove (e);
+
+			CheckPosition (e);
+		}
+
+		protected override void OnMouseLeave (MouseEventArgs e)
+		{
+			base.OnMouseLeave (e);
+			IsMouseOverItem = false;
+		}
+
+		private TreeViewEx topParent;
+
+		private UIElement GetNextElement (bool previous)
+		{
+			if (!previous && HasItems && IsExpanded)
+				return (UIElement)ItemContainerGenerator.ContainerFromIndex (0);
+
+			var parentContainer = this.FindParent<ItemsControl>();
+			if (parentContainer == null)
+				return null;
+
+			int index = parentContainer.ItemContainerGenerator.IndexFromContainer (this);
+			index += (previous) ? -1 : 1;
+
+			if (index == -1)
+				return parentContainer;
+
+			while (index == parentContainer.Items.Count) {
+				if (parentContainer is TreeView)
+					return parentContainer;
+				
+				var parentsParent = parentContainer.FindParent<ItemsControl>();
+				index = parentsParent.ItemContainerGenerator.IndexFromContainer (parentContainer) + 1;
+
+				parentContainer = parentsParent;
+			}
+
+			var item = (TreeViewItem)parentContainer.ItemContainerGenerator.ContainerFromIndex (index);
+			if (!previous || (!item.IsExpanded || !item.HasItems))
+				return item;
+
+			return (UIElement)item.ItemContainerGenerator.ContainerFromIndex (item.Items.Count - 1);
+		}
+
+		private TreeViewEx GetParentTree()
+		{
+			if (this.topParent == null)
+				this.topParent = this.FindParent<TreeViewEx>();
+
+			return this.topParent;
+		}
+
+		private void CheckPosition (MouseEventArgs e)
```

Comment:
Since this is essentially causing `IsMouseOverItem` to be changed, I would not keep that as a side-effect of a void-returning method, but rather have client code set the property explicitly: instead of doing `CheckPosition(mea)`, do a more explicit `IsMouseOverItem = CheckMouseOverThis(mea)`.